p4-16/spec/P4-16-spec.mdk

```diff
@@ -7097,16 +7100,155 @@ annotationBody
     | annotationBody annotationToken
 ~ End P4Grammar
 
-Annotations may impose additional structure on their bodies, and are not
-confined to the P4 language. For example, the P4Runtime specification defines a
-`@pkginfo` annotation that takes key-value pairs, which are not found in the
-main P4 language.
+Unstructured annotations may impose additional structure on their bodies, and
+are not confined to the P4 language. For example, the P4Runtime specification
+defines a `@pkginfo` annotation that takes key-value pairs, which are not found
+in the main P4 language.
+
+## Bodies of Structured Annotations
+
+Unlike unstructured annotations, structured annotations use square brackets
+`[...]` and have a restricted format. They are commonly used to declare custom
+metadata, consisting of expression lists or key-value lists but not both. An
+`expressionList` may be empty or contain a comma-separated list of member
+`expression`s. A `kvList` consists of one or more `kvPair`s, each consisting of
+a key and a value `expression`. Note the syntax for `expression` is rich, see
+Appendix [#sec-grammar] for details.
+
+To avoid ambiguity, an empty `stucturedAnnotationBody` is by definition
+considered to contain an empty `expressionList`.
+
+All `expression`s within a `stucturedAnnotationBody` must evaluate at
+compile-time to one of the following base types: string literals,
```

Comment:
I would phrase this as "... must evaluate to a compile-time constant of the following types: string, infinite-precision integer, or boolean".

Reply:
In another recent discussion, I was reminded that the term that the language spec maintainers would like to use for this is "compile-time known value", so if a change is made here, perhaps a desired change would be to replace this:

"All `expression`s within a `stucturedAnnotationBody` must evaluate at
compile-time to one of the following base types: string literals, infinite-precision integers or booleans."

with this:

"All `expressions within a `stucturedAnnotationBody` must be compile-time known values, either literals or expressions requiring compile-time evaluation, with a result type that is one of: string literals, infinite-precision integer, or boolean."

Reply:
Done, thanks.